codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java

```diff
@@ -495,20 +518,22 @@ private void incrementPendingBytes(int numBytes) {
              * Note: this does not flush the {@link ChannelHandlerContext}.
              */
             int write(int allowedBytes) {
-                int before = payload.size();
-                needFlush |= payload.write(Math.max(0, allowedBytes));
-                int writtenBytes = before - payload.size();
                 try {
+                    pendingWriteQueue.remove();
```

Comment:
@Scottmitch @buchgr 
I think the current separation of responsibility between the encoder and flow controller makes sense in general. The flow controller manages the queue of pending writes but doesn't directly affect the connection state.  The encoder manages the connection state, but doesn't know about the pending writes directly. I think giving the encoder access to the `pendingWriteQueue` violates this separation and would couple the encoder with the flow controller implementation details. We'd be asking all implementations of `payload.write` to implement this logic properly, which doesn't seem reasonable ... it seems like the flow controller should just handle this.

So to back up a bit ... the trouble we're getting into is in how the encoder and flow controller interact. The flow controller doesn't control what is done by `payload.write()`. If the encoder closes/half-closes the stream (due to EOS or an error), this immediately triggers the flow controller to clear the queue for that stream (i.e. error any pending writes still in the queue).  If the flow controller is in the middle of processing the pending write at the head of the queue when this happens, badness ensues (especially in the case of EOS, since we would write the stream and then kill it with RST_STREAM). Whether we add `writeComplete` to the `FlowControlled` interface or not, we still have to deal with the unknown behavior of `payload.write()`.

I think having the flow controller pop the queue before calling `payload.write` is a reasonable work-around.  The one thing that I see is missing in the `finally` logic below is to check that the stream is still writable before restoring the pending write.

WDYT?